src/adc.rs

```diff
@@ -17,6 +11,9 @@ pub use continuous::{
     Attenuated, ChainedAdcChannels, EmptyAdcChannels,
 };
 
+#[cfg(not(feature = "adc-oneshot-new"))]
+pub use oneshot_legacy::*;
```

Comment:
i don't want to bikeshed to hard on naming, though we might want to just call it legacy_adc since there was no hard difference between the modes in the old driver right? 

Furthermore in the special case of the adc driver, since the new driver has both a oneshot and a continuous mode we would need to gate the legacy driver behind two features, or we could just simply create a general adc-new feature that activates both the oneshot and the continouse api as that would not create a problem?

Reply:
> i don't want to bikeshed to hard on naming, though we might want to just call it legacy_adc since there was no hard difference between the modes in the old driver right?

No. There is a hard difference.
The legacy continuous ADC driver (which we never wrapped) is one thing. The legacy oneshot is another.
Legacy oneshot conflicts with new oneshot only. Legacy continuous conflicts with the new continuous only. But no cross-conflicts (or that's how I read the ESP IDF C code). 

But let's even not have a reason for "bikeshedding" in the first place. I did the `oneshot_legacy` `pub` by mistake. We probably don't want this (yet) as that would be a new API. Only the re-export stays in now, the `mod` itself is now private.

> Furthermore in the special case of the adc driver, since the new driver has both a oneshot and a continuous mode we would need to gate the legacy driver behind two features, or we could just simply create a general adc-new feature that activates both the oneshot and the continouse api as that would not create a problem?

The new ADC **component** has two separate drivers: one for "oneshot", and then another - for "continuous". Other than the fact that these two drivers are (for whatever reason) in a single component, they are otherwise two separate drivers, just like the legacy ones. And the legacy ones have two separate "CONFLICT!" checks, as per above.

Reply:
To be clear i am talking about the C drivers here. The old adc driver lived just in [driver/adc.h](https://github.com/espressif/esp-idf/blob/v4.4.7/components/driver/include/driver/adc.h). While it had a configuration for "singel read" that doesn't mean that the driver at that time had any notion of "oneshot" or "continouse". Retroactively fitting names to that driver doesn't make sense. 

Do i miss something here or where do you see that "legacy continuous" definition that we don't wrap  and "legacy oneshot" ?  In the case of the "legacy continuous" are you talking about the the "[adc_dma_legacy](https://github.com/espressif/esp-idf/blob/8760e6d2a7e19913bc40675dd71f374bcd51b0ae/components/driver/deprecated/adc_dma_legacy.c#L625)" stuff`? When was that introduced anyway, i think it was not part of idf v4? was it only briefly in v5.0 or something?

Anyway currently we only have an "old adc" driver implementation utilizing the "legacy_adc" and  that probably will not change anymore since for example it doesn't even work  on most riscv chips correctly. So the goal is to disable the old driver when the new drivers are active. So we only have to worry about that check right? In that light i think it would be fine to just call the feature "adc-new"

Reply:
Yes, "dma" = "digi" = "continuous", and no it did not disappear, it is right there in front of you in the 4.4.7 documentation you are citing (look for methods containing `_digi_` in their names).

While you are right, that there is a common header for both the oneshot and continuous legacy drivers (`adc.h`) they are two legacy drivers, in that they do not share configuration or any common infrastructure. In the 4.4.7 documentation you are looking:
- adc_digi_* is the "continuous (**=DMA**) driver
- everything "non-digi" is the "oneshot" (= non-DMA) driver

By the way, I'm not sure the "digi" (= continuous) one was documented before some 4.4.X point-release, but now it is. 

In 5.x they changed the terminology in that "digi" is now named "continuous" as "digi" was indeed weird. Everything is "digi" with computers, after all.

As for "oneshot" being a stupid name, or out of place etc: what would you call the non-continuous non-DMA driver? We need a name. Hence - oneshot, as the e-hal calls it. And in fact, _even ESP IDF calls it that way if you look_ **at the first link below**.

As for why I think oneshot and "digi" (= continuous) are two _separate_ drivers, even if they share a common component (in the new driver) and a common header (in the legacy sources), that's because ESP IDF seems to treat them as such.
Look [here](https://github.com/espressif/esp-idf/blob/master/components/driver/deprecated/adc_legacy.c#L919)
... and then [here](https://github.com/espressif/esp-idf/blob/master/components/driver/deprecated/adc_dma_legacy.c#L631)

If "oneshot" and "continuous" are not two separate drivers, why are there TWO checks, instead of just one check?

Reply:
> Furthermore in the special case of the adc driver, since the new driver has both a oneshot and a continuous mode we would need to gate the legacy driver behind two features, or we could just simply create a general adc-new feature that activates both the oneshot and the continouse api as that would not create a problem?

My point is, I don't see why you can't use the legacy "oneshot" driver with the new "continuous" driver? Only using legacy "oneshot" with new "oneshot" is the problem?

Reply:
> The old adc driver lived just in [driver/adc.h](https://github.com/espressif/esp-idf/blob/v4.4.7/components/driver/include/driver/adc.h). 

I don't know whether that would help, but I'll repeat once again: headers don't matter. What matters is the .`c` files, as they result in `.o` files. If legacy "digi" and legacy "oneshot" live in two separate `.c` files (as seems to be the case), then we can't have a conflict by using legacy digi/continuous with new oneshot, or the other way around. legacy "digi" and new "digi" or legacy "oneshot" and new "oneshot" seems to be the only conflict.

Reply:
I don't think that the current "oneshot" name is bad. To me it was just not clear what oneshot_legacy was referring too e.g that we say that's the old adc.c and whatever had a "digi"/dma part in it is the equivalent of the continuous case. So spelling the connection out lout helps avoid confusion.

Reply:
OK but **what action do you suggest/agree/disagree with here**? Name the new feature "adc-legacy" because you still don't agree with me that oneshot and continuous are two separate drivers in both 4.4.X and 5.X, or name the new feature "adc-legacy-oneshot" as you finally agree that these are two separate drivers?

Reply:
I agree to you that we have two seperate .c api's both in old and new cases. The mod name "oneshot_legacy" was just not clear to me intuitively because previously i did not distinguished between the "digital" and "adc.c" api. That we cleared uo. Adding a small comment like" // oneshot_legacy is referring to the deprecated adc.c part of the adc.h api inside esp-idf" might help or people will just see this PR so it also might be fine. 

We only export one legacy driver - the onesho_legacyt one. So shortening the feature name from "adc-oneshot-legacy" to "adc-legacy" would be ok in my eyes. But that is not important to me and i would also be fine if you think its more intuitive to call it "adc-oneshot-legacy" or something else. 

My key point: We have a breaking change in either way: oneshot or oneshot_legacy because of the feature gate. Removing a cfg(not(feature)) is not a breaking change in the future while removing a cfg(feature) is. So adding a cfg(feature) to legacy is only breaking once - while adding to oneshot is potentially breaking twice( once now and once in the future). 

And since we are now talking about gating the legacy by default i would like to call the feature "adc-legacy" or "adc-oneshot-legacy" than "adc-new".

Reply:
I hope i did not oneshot you, still thank you for putting up with me and do the initial work here ;D